feat: tap to see next/previous image

thezeroalpha opened the pull request:
## Description
This change enables switching to the next/previous photo in the photo viewer by tapping the left/right quarter of the screen respectively. Videos are not affected.
It's implemented following the feature request in [this discussion](https://github.com/immich-app/immich/discussions/8155).
Tapping in the middle area of the screen still shows/hides controls as before.

## How Has This Been Tested?

I tested it manually on my phone (Samsung Galaxy S10, beyond1lte) with two Immich servers.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable (no applicable changes)
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary. (no new dependencies)
- [x] I have written tests for new code (if applicable) (no applicable automatic tests)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

thezeroalpha committed:
feat(mobile): tap behavior for next/previous image

This change enables switching to the next/previous photo in the photo
viewer by tapping the left/right quarter of the screen.

shenlong-tanwen commented on mobile/lib/pages/common/gallery_viewer.page.dart:
```diff
@@ -244,8 +244,24 @@ class GalleryViewerPage extends HookConsumerWidget {
+
+          int? currentPage = controller.page?.toInt();
+
+          if (tappedLeftSide && currentPage != null) {
+              controller.jumpToPage(currentPage - 1);
+          } else if (tappedRightSide && currentPage != null) {
+            controller.jumpToPage(currentPage + 1);
```
Check if we are at the very first asset or the very last asset before jumping. If not, there is a weird animation when you tap from the first or last asset.

shenlong-tanwen commented:
Thanks a lot for working on this. We are actively migrating away from Isar to Sqlite. `asset_viewer.page.dart` is the new gallery viewer. Can you make your changes over there as well?

thezeroalpha committed:
Avoid animation on first/last image

thezeroalpha committed:
Add changes to asset_viewer.page

thezeroalpha commented:
Yes, that's a valid concern which also came up in https://github.com/immich-app/immich/discussions/8155. I had originally thought to disable this behavior by default, and add a settings toggle to enable it. However, the feedback was that this is unnecessary, also because there are other apps with the same behavior.

I don't have a strong preference; actually, I think when I started using Immich, I expected tapping to work for navigation. But I can implement it either way -- not sure what others think.

1661410107114145152525 commented:
> Thank you for the pr. I have a concern about the usability of the PR. The action is quite unexpected from the user pov. Usually in the asset viewer, tapping on the screen is to hide/show the control bars, not for navigation.
> 
> I think it will confuse the users, what do you think?

As this is meanwhile the default for many popular apps like Google Maps, Instagram Stories or even WhatsApp Stories etc. so intuitivity in General shouldn't be an issue. Tapping at the center should still bring up those bars. 

Only thing this limits is the area where double tap to zoom works but I use it since quite a while in F-stop and just double tap more in the center and then move a bit. Depending on the age/tech affinity of most of your users though older people might accidentially skip more often at the beginning.

It further epends on how large the area to skip to the previous or next image is. I used some css overrides to achieve the same for the web app on mobile and it works like a charm. If I hand my smartphone my grandmother she accidentially forwards to another image sometimes but as swiping still works at the same time she was able to recover quickly.

That said I also understand if people want it to be something with a setting to disable it. If this is needed for a v1 (if this switch is a lot of effort, if not maybe someone could pinpoint @thezeroalpha where/how to add it?) or if it could be added quickly in a v2 if issued really arise is something I can't tell/decide.

YarosMallorca commented:
Please disable this by default; this is very unexpected in photo-viewing apps. I usually tap anywhere on the screen to bring up the overlay menus, and switching to another asset like this is kinda weird to me.

This is a great feature for people who want it, but make them enable this feature in settings.

thezeroalpha committed:
Add setting for tap navigation, disable by default

Not everyone wants to have tapping for next/previous image enabled, so
this commit adds a settings toggle. Since it might be confusing behavior
for new users, it is disabled by default.

thezeroalpha commented:
Thanks for your feedback, the preference is clear. I've added a toggle in the asset viewer settings, and disabled the behavior by default. I've also updated the new gallery viewer. Everything seems to work as expected on my phone. Please let me know if I missed anything, or if I should make further improvements.

leleogere commented:
Hi! I'm the original author of https://github.com/immich-app/immich/discussions/8155. I'm very happy to see someone working on this, and I can't wait to see it integrated into Immich!

I do agree with other points about this *not* being the default behaviour. While it is true that a lot of social media platforms use left/right taps to navigate through story-like carousels, it is absolutely not intuitive for a gallery to behave like that (the only app I know proposing this feature is F-Stop, and it's not even enabled by default). I'm afraid enabling it by default would cause quite a lot of confusion.

This does not bring much to the discussion, as the current implementation is already disabled by default, but as @alextran1502 wanted to gather some feedback I thought I could still add my view.

jrasm91 commented:
@alextran1502 what's the status of this PR?

BetaRavener commented:
@shenlong-tanwen it seems the PR was updated to work with both asset viewers as requested, along with settings to toggle the behavior. Could this be now merged?

I'm another user that would appreciate such feature, as I find it ergonomically easier to just tap instead of swiping. Another app that supports this is galery view in X-plore file browser. So it's not just IG-style apps.

thezeroalpha commented:
I rebased and re-tested the changes as of https://github.com/immich-app/immich/pull/20286#issuecomment-3127820609 on my own device, everything still seems to work as expected.

timonrieger reviewed (commented):
tested on iOS, navigation and setting working as expected!

bwees commented:
@alextran1502 will have final say on this so we will wait for an approval from him.

alextran1502 committed:
Merge remote-tracking branch 'origin/main' into mobile/tap-to-next-previous-image

# Conflicts:
#	mobile/lib/presentation/widgets/asset_viewer/asset_viewer.page.dart

alextran1502 committed:
chore: refactor

alextran1502 committed:
fix: lint
